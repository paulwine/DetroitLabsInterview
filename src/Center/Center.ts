import React from "react";
import { ForecastDay } from "../Models/ForecastDay";


export enum ActiveView {
    CurrentTemp,
    FiveDay
}

interface CenterState {
    activeView: ActiveView;
    userLocation?: {
        latitude: number;
        longitude: number;
    }; 
    userCity?: string;
    currentTemp : number;
    forecastDays: ForecastDay[];
}

const apiKey : string = '475427f62c3d25e731ed6a19c1e83f92';
const apiEndpoint : string = 'https://api.openweathermap.org/data/2.5/';
const iconEndpoint = "http://openweathermap.org/img/w/";

export class Center {
    private static instance: Center;
    public state : CenterState;

    private constructor() { 
        this.state = {
            activeView: ActiveView.CurrentTemp,
            currentTemp: 0,
            forecastDays: []
        }
    }

    public static center(): Center {
        if (!Center.instance) {
            Center.instance = new Center();
        }
        return Center.instance;
    }

    // ================== Update Functions ============================
    private listeners: React.Component[] = [];

    public static addListener(listener: React.Component) {
        Center.center().listeners.push(listener);
    }

    public static removeListener(listener: React.Component) {
        const index = Center.center().listeners.indexOf(listener);
        if (index > -1) {
            Center.center().listeners.splice(index, 1);
        }
    }

    public static update(): void {
        Center.center().listeners.forEach(listener => listener.setState({}));
    }

    //================= UI Functions =====================================

    public static navigate(view: ActiveView) : void {
        Center.center().state.activeView = view;
        Center.update();
    }

    //================== Data Functions =====================================

    public static setUserLocation() {
        navigator.geolocation.getCurrentPosition((position) => {
            let coords = {
                latitude: position.coords.latitude,
                longitude: position.coords.longitude
            }
            Center.center().state.userLocation = coords;
            Center.update();
            Center.getWeatherData();
        })

    }

    public static async getWeatherData() : Promise<void> {
        const state = Center.center().state;
        if (!state.userLocation){
            console.error("No User Location Set");
            return;
        }
        const lat = state.userLocation.latitude;
        const lon = state.userLocation.longitude;

        const currentTempEndpoint = apiEndpoint + 
        `weather?lat=${lat}&lon=${lon}&appid=${apiKey}&units=imperial`;

        const currentTempResponse = await fetch(currentTempEndpoint);
        const currentTempJSON = await currentTempResponse.json();
        const currentTemp = currentTempJSON["main"]["temp"];

        state.currentTemp = currentTemp;
        state.userCity = currentTempJSON["name"];

        const fiveDayEndpoint = apiEndpoint + 
        `forecast?lat=${lat}&lon=${lon}&appid=${apiKey}&units=imperial`;

        const fiveDayResponse = await fetch(fiveDayEndpoint);
        const fiveDayJSON = await fiveDayResponse.json();

        const list = fiveDayJSON["list"];

        let forecastDays: ForecastDay[] = [];

        list.forEach((item : any) => {
            const temperature : number = item["main"]["temp"];
            const imgUrl : string = iconEndpoint+ item["weather"][0]["icon"] + '.png';
            const date : Date = new Date(item["dt_txt"]);

            const forecastDay : ForecastDay = {
                imgUrl,
                temperature,
                date
            };
            forecastDays.push(forecastDay);
        })

        state.forecastDays = forecastDays;

        Center.update();
    }
}