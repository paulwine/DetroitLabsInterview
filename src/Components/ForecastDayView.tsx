import { Card, Typography } from "@material-ui/core";
import React from "react";
import { ForecastDay } from "../Models/ForecastDay";

interface ForecastDayViewProps {
    items: ForecastDay[];
}

export class ForecastDayView extends React.Component<ForecastDayViewProps>{
    constructor(props: ForecastDayViewProps){
        super(props);
    }
    render() {
        if (!this.props.items || this.props.items.length == 0){
            return (
                <div>
                    <h2>User Location Not Set</h2>
                </div>
            )
        }
        return (
            <div style={containerStyle}>
            <Typography variant={"h5"} style={{color: 'white'}}> {this.props.items[0].date.toLocaleDateString()}</Typography>
                {this.props.items.map(day =>{
                    return (
                        <Card style={itemStyle} elevation={3}>
                            <h5>{day.date.toLocaleTimeString()}</h5>
                            <img src={day.imgUrl}></img>
                            <h4>{day.temperature}Â°F</h4> 
                        </Card>
                    )
                })}
            </div>
        )
    }
}

const containerStyle = {
    background: '#4287f5',
    padding: 10
}

const itemStyle = {
    margin: 10,
    padding: 10,
    display: 'flex',
    justifyContent: 'space-between',
    border: '2px solid gray'

}