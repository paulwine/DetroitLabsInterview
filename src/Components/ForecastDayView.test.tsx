import 'regenerator-runtime/runtime'
import React from 'react'
import {Center, ActiveView } from '../Center/Center'
import {shallow, configure } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16';
import 'jest-fetch-mock';
import { ForecastDayView } from './ForecastDayView';

configure({ adapter: new Adapter() });

describe('ForecastDayView', () =>{
    it('should render forecastDays correctly' , async () => {
        const forecastDays =[
            {"imgUrl":"http://openweathermap.org/img/w/10d.png","temperature":82.4,"date": new Date("2021-06-13T04:00:00.000Z")},
            {"imgUrl":"http://openweathermap.org/img/w/10n.png","temperature":74.62,"date": new Date("2021-06-13T07:00:00.000Z")},
            {"imgUrl":"http://openweathermap.org/img/w/10n.png","temperature":67.96,"date": new Date("2021-06-13T10:00:00.000Z")},
            {"imgUrl":"http://openweathermap.org/img/w/10n.png","temperature":67.6,"date": new Date("2021-06-13T13:00:00.000Z")},
            {"imgUrl":"http://openweathermap.org/img/w/03d.png","temperature":67.57,"date": new Date("2021-06-13T16:00:00.000Z")},
            {"imgUrl":"http://openweathermap.org/img/w/04d.png","temperature":71.83,"date": new Date("2021-06-13T19:00:00.000Z")},
            {"imgUrl":"http://openweathermap.org/img/w/03d.png","temperature":80.17,"date": new Date("2021-06-13T22:00:00.000Z")},
            {"imgUrl":"http://openweathermap.org/img/w/01d.png","temperature":80.83,"date": new Date("2021-06-14T01:00:00.000Z")}
        ];
        const forecastDayView = shallow(<ForecastDayView items={forecastDays}/>)
        expect(forecastDayView.html()).toMatchSnapshot();
    })
    
})