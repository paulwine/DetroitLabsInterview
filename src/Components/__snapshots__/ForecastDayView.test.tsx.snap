// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ForecastDayView should render forecastDays correctly 1`] = `"<div style=\\"background:#4287f5;padding:10px\\"><h5 class=\\"MuiTypography-root MuiTypography-h5\\" style=\\"color:white\\"> 6/13/2021</h5><div class=\\"MuiPaper-root MuiCard-root MuiPaper-elevation3 MuiPaper-rounded\\" style=\\"margin:10px;padding:10px;display:flex;justify-content:space-between;border:2px solid gray\\"><h5>12:00:00 AM</h5><img src=\\"http://openweathermap.org/img/w/10d.png\\"/><h4>82.4°F</h4></div><div class=\\"MuiPaper-root MuiCard-root MuiPaper-elevation3 MuiPaper-rounded\\" style=\\"margin:10px;padding:10px;display:flex;justify-content:space-between;border:2px solid gray\\"><h5>3:00:00 AM</h5><img src=\\"http://openweathermap.org/img/w/10n.png\\"/><h4>74.62°F</h4></div><div class=\\"MuiPaper-root MuiCard-root MuiPaper-elevation3 MuiPaper-rounded\\" style=\\"margin:10px;padding:10px;display:flex;justify-content:space-between;border:2px solid gray\\"><h5>6:00:00 AM</h5><img src=\\"http://openweathermap.org/img/w/10n.png\\"/><h4>67.96°F</h4></div><div class=\\"MuiPaper-root MuiCard-root MuiPaper-elevation3 MuiPaper-rounded\\" style=\\"margin:10px;padding:10px;display:flex;justify-content:space-between;border:2px solid gray\\"><h5>9:00:00 AM</h5><img src=\\"http://openweathermap.org/img/w/10n.png\\"/><h4>67.6°F</h4></div><div class=\\"MuiPaper-root MuiCard-root MuiPaper-elevation3 MuiPaper-rounded\\" style=\\"margin:10px;padding:10px;display:flex;justify-content:space-between;border:2px solid gray\\"><h5>12:00:00 PM</h5><img src=\\"http://openweathermap.org/img/w/03d.png\\"/><h4>67.57°F</h4></div><div class=\\"MuiPaper-root MuiCard-root MuiPaper-elevation3 MuiPaper-rounded\\" style=\\"margin:10px;padding:10px;display:flex;justify-content:space-between;border:2px solid gray\\"><h5>3:00:00 PM</h5><img src=\\"http://openweathermap.org/img/w/04d.png\\"/><h4>71.83°F</h4></div><div class=\\"MuiPaper-root MuiCard-root MuiPaper-elevation3 MuiPaper-rounded\\" style=\\"margin:10px;padding:10px;display:flex;justify-content:space-between;border:2px solid gray\\"><h5>6:00:00 PM</h5><img src=\\"http://openweathermap.org/img/w/03d.png\\"/><h4>80.17°F</h4></div><div class=\\"MuiPaper-root MuiCard-root MuiPaper-elevation3 MuiPaper-rounded\\" style=\\"margin:10px;padding:10px;display:flex;justify-content:space-between;border:2px solid gray\\"><h5>9:00:00 PM</h5><img src=\\"http://openweathermap.org/img/w/01d.png\\"/><h4>80.83°F</h4></div></div>"`;
